cluster/gce/upgrade.sh

```diff
@@ -343,51 +343,61 @@ function do-node-upgrade() {
-    echo "== Calling rolling-update for ${group}. ==" >&2
-    update=$(gcloud alpha compute rolling-updates \
+    gcloud compute instance-groups managed set-instance-template "${group}" \
+      --template="${template_name}" \
+      --project="${PROJECT}" \
+      --zone="${ZONE}"
+    set_instance_template_rc=$?
+    if [[ "${set_instance_template_rc}" != 0 ]]; then
+      return ${set_instance_template_rc}
+    fi
+    gcloud compute instance-groups managed list-instances "${group}" \
+        --format='value(instance)' \
         --project="${PROJECT}" \
         --zone="${ZONE}" \
-        start \
-        --group="${group}" \
-        --template="${template_name}" \
-        --instance-startup-timeout=300s \
-        --max-num-concurrent-instances=1 \
-        --max-num-failed-instances=0 \
-        --min-instance-update-time=0s 2>&1) && update_rc=$? || update_rc=$?
-
-    if [[ "${update_rc}" != 0 ]]; then
-      echo "== FAILED to start rolling-update: =="
-      echo "${update}"
-      echo "  This may be due to a preexisting rolling-update;"
-      echo "  see https://github.com/kubernetes/kubernetes/issues/33113 for details."
-      echo "  All rolling-updates in project ${PROJECT} zone ${ZONE}:"
-      gcloud alpha compute rolling-updates \
+        | while read instance; do
+      echo "== Draining ${instance}. == " >&2
+      "${KUBE_ROOT}/cluster/kubectl.sh" drain --delete-local-data --force --ignore-daemonsets "${instance}"
+      drain_rc=$?
+      if [[ "${drain_rc}" != 0 ]]; then
+        echo "== FAILED to drain ${instance} =="
+        return ${drain_rc}
+      fi
+      gcloud compute instance-groups managed recreate-instances "${group}" \
         --project="${PROJECT}" \
         --zone="${ZONE}" \
-        list || true
-      return ${update_rc}
-    fi
-
-    id=$(echo "${update}" | grep "Started" | cut -d '/' -f 11 | cut -d ']' -f 1)
-    updates+=("${id}")
-  done
-
-  echo "== Waiting for Upgrading nodes to be finished. ==" >&2
-  # Wait until rolling updates are finished.
-  for update in ${updates[@]}; do
-    while true; do
-      result=$(gcloud alpha compute rolling-updates \
-          --project="${PROJECT}" \
-          --zone="${ZONE}" \
-          describe \
-          ${update} \
-          --format='value(status)' || true)
-      if [ $result = "ROLLED_OUT" ]; then
-        echo "Rolling update ${update} is ${result} state and finished."
-        break
+        --instances="${instance}"
+      recreate_rc=$?
+      if [[ "${recreate_rc}" != 0 ]]; then
+        echo "== FAILED to recreate ${instance} =="
+        return ${recreate_rc}
       fi
-      echo "Rolling update ${update} is still in ${result} state."
-      sleep 10
+      echo "== Waiting for ${instance} to become young. ==" >&2
+      while true; do
+        age=$(( $(date +%s) - $(date --date=$("${KUBE_ROOT}/cluster/kubectl.sh"  get node "${instance}" --output='jsonpath={.metadata.creationTimestamp}' ) +%s) ))
```

Comment:
You can watch for the `UID/externalID` of the node object to change. Or you can keep fetching the instance from GCE and wait for the ID to change. This is the authoritative way to know that the instance did indeed get recreated.
Once the ID changes, you need to wait for the node to enter `Ready` state.

`gcloud compute instances describe $INSTANCE --format='get(id)'`